go/libraries/doltcore/sqle/writer/prolly_table_writer.go

```diff
@@ -0,0 +1,296 @@
+	// todo(andy)
+	return nil
+}
+
+// StatementComplete implements TableWriter.
+func (w *prollyWriter) StatementComplete(ctx *sql.Context) error {
+	// todo(andy)
+	return nil
+}
+
+func (w *prollyWriter) table(ctx context.Context) (t *doltdb.Table, err error) {
+	m, err := w.mut.Map(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	t, err = w.tbl.UpdateRows(ctx, durable.IndexFromProllyMap(m))
+	if err != nil {
+		return nil, err
+	}
+
+	if w.aiCol.AutoIncrement {
+		seq, err := w.aiTracker.Next(w.tableName, nil, nil)
+		if err != nil {
+			return nil, err
+		}
+		vrw := w.tbl.ValueReadWriter()
+
+		v, err := w.aiCol.TypeInfo.ConvertValueToNomsValue(ctx, vrw, seq)
+		if err != nil {
+			return nil, err
+		}
+
+		t, err = t.SetAutoIncrementValue(ctx, v)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return t, nil
+}
+
+func (w *prollyWriter) flush(ctx *sql.Context) error {
+	newRoot, err := w.sess.Flush(ctx)
+	if err != nil {
+		return err
+	}
+
+	return w.setter(ctx, w.dbName, newRoot)
+}
+
+type prollyIndexWriter struct {
+	mut prolly.MutableMap
+
+	keyBld *val.TupleBuilder
+	keyMap colMapping
+
+	valBld *val.TupleBuilder
+	valMap colMapping
+}
+
+func makeMutableProllyIndex(m prolly.Map, sqlSch sql.Schema, sch schema.Schema) prollyIndexWriter {
+	keyDesc, valDesc := m.Descriptors()
+	keyMap, valMap := colMappingsFromSchema(sqlSch, sch)
+
+	return prollyIndexWriter{
+		mut:    m.Mutate(),
+		keyBld: val.NewTupleBuilder(keyDesc),
+		keyMap: keyMap,
+		valBld: val.NewTupleBuilder(valDesc),
+		valMap: valMap,
+	}
+}
+
+var sharePool = pool.NewBuffPool()
+
+func (m prollyIndexWriter) Map(ctx context.Context) (prolly.Map, error) {
+	return m.mut.Map(ctx)
+}
+
+func (m prollyIndexWriter) Insert(ctx *sql.Context, sqlRow sql.Row) error {
+	for to, from := range m.keyMap {
+		m.keyBld.PutField(to, sqlRow[from])
+	}
+	k := m.keyBld.Build(sharePool)
+
+	ok, err := m.mut.Has(ctx, k)
+	if err != nil {
+		return err
+	} else if ok {
+		return m.primaryKeyError(ctx, k)
+	}
+
+	for to, from := range m.valMap {
+		m.valBld.PutField(to, sqlRow[from])
+	}
+	v := m.valBld.Build(sharePool)
+
+	return m.mut.Put(ctx, k, v)
+}
+
+func (m prollyIndexWriter) Delete(ctx *sql.Context, sqlRow sql.Row) error {
+	for to, from := range m.keyMap {
+		m.keyBld.PutField(to, sqlRow[from])
+	}
+	k := m.keyBld.Build(sharePool)
+
+	return m.mut.Delete(ctx, k)
+}
+
+func (m prollyIndexWriter) Update(ctx *sql.Context, oldRow sql.Row, newRow sql.Row) error {
+	for to, from := range m.keyMap {
+		m.keyBld.PutField(to, oldRow[from])
+	}
+	oldKey := m.keyBld.Build(sharePool)
+
+	// todo(andy): we can skip building, deleting |oldKey|
+	//  if we know the key fields are unchanged
+	if err := m.mut.Delete(ctx, oldKey); err != nil {
+		return err
+	}
+
+	for to, from := range m.keyMap {
+		m.keyBld.PutField(to, newRow[from])
+	}
+	newKey := m.keyBld.Build(sharePool)
+
+	ok, err := m.mut.Has(ctx, newKey)
```

Comment:
This tracks unapplied mutations?

Reply:
yeah buddy, got a sweet skip list in there